### PR TITLE
Miscellaneous bug fixes

### LINKED ISSUE #30876: Exception to rule in static constructor docs
> "A static constructor runs before an instance constructor."

There is an important exception to this rule.  If a static initializer creates an instance of the class (e.g. to implement the Singleton pattern), the instance constructor will be invoked before the static constructor. There are two competing rules (initalizer before static constructor) and (static constructor before instance constructor), but it's impossible to satisfy both rules.

Test code
```
public class A
{
    public static void Main()
    {
        Debug.Log("A.Main");
        B.DoSomething();
    }
}

public class B
{
    private static B Instance = new B();
    private static int Foo = 42;

    static B()
    {
        Debug.Log("B Static Constructor");
        LoadStuff();
    }

    public B()
    {
        Debug.Log("B Instance Constructor");
        UseStuff();
    }

    public static void DoSomething()
    {
        Debug.Log("B.DoSomething");
    }

    private static void LoadStuff()
    {
        Debug.Log("B.LoadStuff");
    }

    private static void UseStuff()
    {
        Debug.Log("B.UseStuff");
    }
}
```

Log reveals "B Instance Constructor" before "B Static Constructor".  More importantly "B.UseStuff" is called *before* "B.LoadStuff", which has important ramifications.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: aa3ed1c5-1f79-9234-beec-b5771c3fb3c0
* Version Independent ID: ac4f6361-70a7-fb43-101e-c12044f60896
* Content: [Static Constructors - C# Programming Guide](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/static-constructors)
* Content Source: [docs/csharp/programming-guide/classes-and-structs/static-constructors.md](https://github.com/dotnet/docs/blob/main/docs/csharp/programming-guide/classes-and-structs/static-constructors.md)
* Product: **dotnet-csharp**
* Technology: **csharp-fundamentals**
* GitHub Login: @BillWagner
* Microsoft Alias: **wiwagn**

---
[Associated WorkItem - 291898](https://dev.azure.com/msft-skilling/Content/_workitems/edit/291898)

### LINKED ISSUE #40808: Raw string literal comparison no longer valid
### Describe the issue or suggestion

Viewing [this section on raw literal strings](https://learn.microsoft.com/en-us/dotnet/csharp/programming-guide/strings/#raw-string-literals) , the following text near the bottom of the section is no longer valid because the JSON deserialization article appears to have been updated to make use of raw string literals.

> Compare that text with the equivalent text in the sample on [JSON deserialization](https://learn.microsoft.com/en-us/dotnet/standard/serialization/system-text-json/deserialization), which doesn't make use of this new feature.

To address this, an example could be added to replace the need to view another page, or the paragraph should be removed entirely.

---
#### Document Details

⚠ *Do not edit this section. It is required for learn.microsoft.com ➟ GitHub issue linking.*

* ID: c7aa6134-64e2-28a7-f7d0-d0c1512f4520
* Version Independent ID: 3317fd4a-9cd3-57cd-66f7-abb271e302e8
* Content: [Strings - C# Programming Guide - C#](https://learn.microsoft.com/en-us/dotnet/csharp/programming-guide/strings/#raw-string-literals)
* Content Source: [docs/csharp/programming-guide/strings/index.md](https://github.com/dotnet/docs/blob/main/docs/csharp/programming-guide/strings/index.md)
* Service: **dotnet-csharp**
* Sub-service: **fundamentals**
* GitHub Login: @BillWagner
* Microsoft Alias: **wiwagn**

---
[Associated WorkItem - 291899](https://dev.azure.com/msft-skilling/Content/_workitems/edit/291899)

### LINKED ISSUE #41491: Paragraph about avoiding overriding the same method in multiple derived interfaces is unclear
### Type of issue

Missing information

### Description

In the second to last paragraph of the article, it states: 

> "if both the IBlinkingLight and ITimerLight implemented an override of PowerStatus, the OverheadLight would need to provide a more specific override. Otherwise, the compiler can't pick between the implementations in the two derived interfaces."

But it is unclear how the compiler would be confused here. If the OverheadLight class doesn't provide a more specific override, then the ways to access the PowerStatus from an instance of the OverheadLight class would be to cast or convert it to either an ILight, an IBlinkingLight or an ITimerLight. In those cases, the compiler would know which implementation to pick. Is there another way an instance of OverheadLight could be used to access the PowerStatus that would cause a compiler error?

### Page URL

https://learn.microsoft.com/en-us/dotnet/csharp/advanced-topics/interface-implementation/mixins-with-default-interface-methods

### Content source URL

https://github.com/dotnet/docs/blob/main/docs/csharp/advanced-topics/interface-implementation/mixins-with-default-interface-methods.md

### Document Version Independent Id

98229926-6c7a-3ad1-ab5c-86231253986c

### Article author

@BillWagner

### Metadata

* ID: 2c7899e8-3f01-41d2-137a-e334b73699a4 
* Service: **dotnet-csharp**
* Sub-service: **advanced-concepts**

---
[Associated WorkItem - 269279](https://dev.azure.com/msft-skilling/Content/_workitems/edit/269279)

### LINKED ISSUE #41507: interpolated raw string literal clarifications
### Type of issue

Missing information

### Description

Anonymous verbatim:

In the section on string interpolation, the text describes how to specify the number of `{` and `}` required to denote an interpolated value (multiple `$` characters ) and says that any sequences with FEWER than this number are taken as literal `{` or `}`. 

However, the example given has sequences with MORE than the specified number of `{` or `}` and there is no indication about how this is interpreted (other than an inference from the example output). So, for an interpolated raw string with brace count of 2 (prepended `$$`), how is the sequence `{{{` interpreted?as `{{` `{` or `{` `{{`.  

A similar issue exists in `String.Format` (under .NET Framework). In the example, `{{ {X }}` would be a syntax error (`{X` is not a valid expression). Does the compiler:

a) try various possibilities and choose the (singular) valid one, or
b) treat any sequence of MORE (x) than the required number (n) of `{` as a sequence of (x-n) literal `{` followed by an interpolation, using up the LAST n `{` (or for `}`, first n then (x-n) literal } )

### Page URL

https://learn.microsoft.com/en-us/dotnet/csharp/programming-guide/strings/

### Content source URL

https://github.com/dotnet/docs/blob/main/docs/csharp/programming-guide/strings/index.md

### Document Version Independent Id

3317fd4a-9cd3-57cd-66f7-abb271e302e8

### Article author

@BillWagner

### Metadata

* ID: c7aa6134-64e2-28a7-f7d0-d0c1512f4520 
* Service: **dotnet-csharp**
* Sub-service: **fundamentals**

---
[Associated WorkItem - 269278](https://dev.azure.com/msft-skilling/Content/_workitems/edit/269278)

### LINKED ISSUE #41538: Add collection expressions to "Other usages of []"
### Type of issue

Missing information

### Description

Collection expressions are not included in "Other users of []". 

Also, collection expressions are used without comment early in this article while discussing [] for indexing. Consider adding a comment near the first place collection expressions are used in this article, perhaps with a link.

### Page URL

https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/operators/member-access-operators

### Content source URL

https://github.com/dotnet/docs/blob/main/docs/csharp/language-reference/operators/member-access-operators.md

### Document Version Independent Id

682b8c86-e49a-1d6d-8562-c9f919082bdc

### Article author

@pkulikov

### Metadata

* ID: 0a25abe4-2b4e-119e-85e9-af19ed06e8c0 
* Service: **dotnet-csharp**
* Sub-service: **lang-reference**

---
[Associated WorkItem - 273426](https://dev.azure.com/msft-skilling/Content/_workitems/edit/273426)

### LINKED ISSUE #41629: The documentation of the order of static constructors is wrong.
### Type of issue

Other (describe below)

### Description

The following bit seems wrong.

> 4. Base static constructors run. Any static constructors, starting with [Object.Object](https://learn.microsoft.com/en-us/dotnet/api/system.object.-ctor) through each base class to the direct base class.
> 5. The static constructor runs. The static constructor for the type runs.

As I will demonstrate, in cases where the members of the base class is not accessed by the deriving class, the deriving class constructor will run before the base constructor.

```cs
using static System.Console;

class BaseClass {
    static protected int c = 1;
    static BaseClass() => WriteLine($"{c++}. Static   constructor of *BaseClass*    called.");
    public BaseClass() => WriteLine($"{c++}. Instance constructor of *BaseClass*    called.");
}
class DerivedClass : BaseClass {
    static DerivedClass()          => WriteLine($"{c++}. Static   constructor of  DerivedClass  called.");
    public DerivedClass() : base() => WriteLine($"{c++}. Instance constructor of  DerivedClass  called.");
}

public static class C {
    public static int c = 1;
}
class BaseClass2 {
    static BaseClass2() => WriteLine($"{C.c++}. Static   constructor of *BaseClass2*   called.");
    public BaseClass2() => WriteLine($"{C.c++}. Instance constructor of *BaseClass2*   called.");
}
class DerivedClass2 : BaseClass2 {
    static DerivedClass2()          => WriteLine($"{C.c++}. Static   constructor of  DerivedClass2 called.");
    public DerivedClass2() : base() => WriteLine($"{C.c++}. Instance constructor of  DerivedClass2 called.");
}

public class Program {
    public static void Main() {
        var instance = new DerivedClass();
        WriteLine();
        var instance2 = new DerivedClass2();
    }
}
```

The results are different in the two cases:

```
1. Static   constructor of *BaseClass*    called.
2. Static   constructor of  DerivedClass  called.
3. Instance constructor of *BaseClass*    called.
4. Instance constructor of  DerivedClass  called.

1. Static   constructor of  DerivedClass2 called.
2. Static   constructor of *BaseClass2*   called.
3. Instance constructor of *BaseClass2*   called.
4. Instance constructor of  DerivedClass2 called.
```

### Page URL

https://learn.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/static-constructors

### Content source URL

https://github.com/dotnet/docs/blob/main/docs/csharp/programming-guide/classes-and-structs/static-constructors.md

### Document Version Independent Id

ac4f6361-70a7-fb43-101e-c12044f60896

### Article author

@BillWagner

### Metadata

* ID: aa3ed1c5-1f79-9234-beec-b5771c3fb3c0 
* Service: **dotnet-csharp**
* Sub-service: **fundamentals**

---
[Associated WorkItem - 275698](https://dev.azure.com/msft-skilling/Content/_workitems/edit/275698)

### PR DESCRIPTION
**Pull request by BillWagner**
Combined PR for a few small bug fixes:

- Fixes #41491: Clarified the diamond problem with default interface methods. Add an image to illustrate the problem, and provide additional text explanation.
- Fixes #40808: Remove the (now) obsolete paragraph.
- Fixes #41507: Clarify how additional `{` and `}` characters are parsed in interpolated raw string literals.
- Fixes #41538: Mention collection expressions in the member access article. Include links to the collection expressions article.
- Fixes #30876: Mention that when a static field initializer creates an instance of a type, that instance constructor is invoked before the static constructor.
- Fixes #41629: Note that the execution order of static constructors isn't specified. While here, perform an edit pass.

The fixes are in the order of the commits, so reviewing commit-by-commit may be more useful.

---

#### Internal previews

| 📄 File | 🔗 Preview link |
|:--|:--|
| [docs/csharp/advanced-topics/interface-implementation/mixins-with-default-interface-methods.md](https://github.com/dotnet/docs/blob/5c22305f1a796adebf8b0a414734a1da8c82daad/docs/csharp/advanced-topics/interface-implementation/mixins-with-default-interface-methods.md) | [Tutorial: Mix functionality in when creating classes using interfaces with default interface methods](https://review.learn.microsoft.com/en-us/dotnet/csharp/advanced-topics/interface-implementation/mixins-with-default-interface-methods?branch=pr-en-us-41967) |
| [docs/csharp/language-reference/operators/member-access-operators.md](https://github.com/dotnet/docs/blob/5c22305f1a796adebf8b0a414734a1da8c82daad/docs/csharp/language-reference/operators/member-access-operators.md) | [Member access operators and expressions - the dot, indexer, and invocation operators.](https://review.learn.microsoft.com/en-us/dotnet/csharp/language-reference/operators/member-access-operators?branch=pr-en-us-41967) |
| [docs/csharp/language-reference/tokens/interpolated.md](https://github.com/dotnet/docs/blob/5c22305f1a796adebf8b0a414734a1da8c82daad/docs/csharp/language-reference/tokens/interpolated.md) | ["$ - string interpolation - format string output"](https://review.learn.microsoft.com/en-us/dotnet/csharp/language-reference/tokens/interpolated?branch=pr-en-us-41967) |
| [docs/csharp/programming-guide/classes-and-structs/static-constructors.md](https://github.com/dotnet/docs/blob/5c22305f1a796adebf8b0a414734a1da8c82daad/docs/csharp/programming-guide/classes-and-structs/static-constructors.md) | [Static Constructors (C# Programming Guide)](https://review.learn.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/static-constructors?branch=pr-en-us-41967) |
| [docs/csharp/programming-guide/strings/index.md](https://github.com/dotnet/docs/blob/5c22305f1a796adebf8b0a414734a1da8c82daad/docs/csharp/programming-guide/strings/index.md) | [docs/csharp/programming-guide/strings/index](https://review.learn.microsoft.com/en-us/dotnet/csharp/programming-guide/strings/index?branch=pr-en-us-41967) |

### CODE REVIEW

**Comment by BillWagner** [comment]
ping @adegeo for review

File: docs/csharp/programming-guide/strings/index.md [hunk 1]
```diff
@@ -53,7 +53,7 @@ For more information about how to create new strings that are based on modificat
 
 Beginning with C# 11, you can use *raw string literals* to more easily create strings that are multi-line, or use any characters requiring escape sequences. *Raw string literals* remove the need to ever use escape sequences. You can write the string, including whitespace formatting, how you want it to appear in output. A *raw string literal*:
 
-- Starts and ends with a sequence of at least three double quote characters (`"""`). You're allowed more than three consecutive characters to start and end the sequence in order to support string literals that contain three (or more) repeated quote characters.
+- Starts and ends with a sequence of at least three double quote characters (`"""`). You can use more than three consecutive characters to start and end the sequence. You can embed string literals that contain three (or more) repeated quote characters.
```
Comment (adegeo) [inline docs/csharp/programming-guide/strings/index.md:56]: I really feel like the previous explanation was clearer. Effectively these statements are independent now:

- You can use more than three consecutive characters to start and end the sequence.
- You can embed string literals that contain three (or more) repeated quote characters.

While the second statement _is_ true, it's only true (without the 'or more' clause) if you use more than three characters for the sequence. But that's not clear anymore.

**Review by adegeo — APPROVED** [review]
LGTM, one thing to consider.

File: docs/csharp/programming-guide/strings/index.md (hunk 1 above)
Comment (BillWagner) [reply docs/csharp/programming-guide/strings/index.md:56]: Right. I sometimes hate acrolinx. Let's revert:

```suggestion
- Starts and ends with a sequence of at least three double quote characters (`"""`). You can use more than three consecutive characters to start and end the sequence to support string literals that contain three (or more) repeated quote characters.
```

**Pull request merged**